### ISSUE TITLE
Open-Q 820 MSM9886

### ISSUE THREAD

**Issue by rabits** [bkerler/edl#238]
Another better bugreport based on #232 

I figured out how to read the info by `--skipresponse`, but quickly found another issue with write and finally found how to solve it with a dirty hack.

So I thought the firehose just doesn't fit for me and I have to use another one, so returned back to the buggy `Loaders/xiaomi/009470e100000000_355d47f912fea0af_fhprg_peek.bin` and it quickly shows me the same exception:
```
...
main - Trying to connect to firehose loader ...
Traceback (most recent call last):
  File "/usr/local/bin/edl", line 4, in <module>
    __import__('pkg_resources').run_script('edlclient==3.60', 'edl')
  File "/usr/local/lib/python3.8/dist-packages/pkg_resources/__init__.py", line 650, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/usr/local/lib/python3.8/dist-packages/pkg_resources/__init__.py", line 1446, in run_script
    exec(code, namespace, namespace)
  File "/usr/local/lib/python3.8/dist-packages/edlclient-3.60-py3.8.egg/EGG-INFO/scripts/edl", line 380, in <module>
    base.run()
  File "/usr/local/lib/python3.8/dist-packages/edlclient-3.60-py3.8.egg/EGG-INFO/scripts/edl", line 375, in run
    if fh.connect(sahara):
  File "/usr/local/lib/python3.8/dist-packages/edlclient-3.60-py3.8.egg/edlclient/Library/firehose_client.py", line 109, in connect
    if self.firehose.configure(0):
  File "/usr/local/lib/python3.8/dist-packages/edlclient-3.60-py3.8.egg/edlclient/Library/firehose.py", line 851, in configure
    if "MemoryName" in rsp.data:
TypeError: a bytes-like object is required, not 'str'
```

But using this small patch I get the firehose to work for me and write the partition I want:
```patch
diff --git a/edlclient/Library/firehose.py b/edlclient/Library/firehose.py
index fbeaf89..a13c168 100755
--- a/edlclient/Library/firehose.py
+++ b/edlclient/Library/firehose.py
@@ -848,8 +848,12 @@ class firehose(metaclass=LogBase):
         rsp = self.xmlsend(connectcmd)
         if not rsp.resp:
             if rsp.error=="":
-                if "MemoryName" in rsp.data:
-                    self.cfg.MemoryName = rsp.data["MemoryName"]
+                try:
+                    if "MemoryName" in rsp.data:
+                        self.cfg.MemoryName = rsp.data["MemoryName"]
+                except TypeError:
+                    self.warning("!DEBUG! rsp.data: '%s'" % (rsp.data,))
+                    return self.configure(lvl + 1)
                 if "MaxPayloadSizeFromTargetInBytes" in rsp.data:
                     self.cfg.MaxPayloadSizeFromTargetInBytes = int(rsp.data["MaxPayloadSizeFromTargetInBytes"])
                 if "MaxPayloadSizeToTargetInBytes" in rsp.data:
```

Result:
```
$ edl w rpm rpmbak.img  --memory UFS --loader Loaders/xiaomi/009470e100000000_355d47f912fea0af_fhprg_peek.bin
Qualcomm Sahara / Firehose Client V3.60 (c) B.Kerler 2018-2022.
main - Using loader Loaders/xiaomi/009470e100000000_355d47f912fea0af_fhprg_peek.bin ...
main - Waiting for the device
main - Device detected :)
sahara - Protocol version: 2.1
main - Mode detected: sahara
sahara -
------------------------
HWID:              0x0003e0e100000000 (MSM_ID:0x0003e0e1,OEM_ID:0x0000,MODEL_ID:0x0000)
Unknown CPU, please send log as issue to https://github.com/bkerler/edl
PK_HASH:           0x7be49b72f9e4337223ccb84d6eccca4e61ce16e3602ac2008cb18b75babe6d09
Serial:            0x2b5f6219

sahara - Protocol version: 2.1
sahara - Uploading loader Loaders/xiaomi/009470e100000000_355d47f912fea0af_fhprg_peek.bin ...
sahara - 64-Bit mode detected.
sahara - Firehose mode detected, uploading...
sahara - Loader successfully uploaded.
main - Trying to connect to firehose loader ...
firehose
firehose - [LIB]: !DEBUG! rsp.data: 'bytearray(b'<?xml version="1.0" encoding="UTF-8" ?>\n<data>\n<log value="Binary build date: Feb 27 2016 @ 21:09:42"/>\n</data><?xml version="1.0" encoding="UTF
-8" ?>\n<data>\n<log value="Chip serial num: 727671321 (0x2b5f6219)"/>\n</data><?xml version="1.0" encoding="UTF-8" ?>\n<data>\n<log value="Supported Functions: program configure nop firmwarewrite
 patch setbootablestoragedrive ufs emmc power benchmark read getstorageinfo getsha256digest erase peek poke "/>\n</data><?xml version="1.0" encoding="UTF-8" ?>\n<data>\n<log value="ufs 198,3b9e000
"/>\n</data>')'
firehose - TargetName=MSM8996
firehose - MemoryName=UFS
firehose - Version=1
firehose - Trying to read first storage sector...
firehose - Running configure...
firehose - {'Device Total Logical Blocks': '0x60ac00', 'Device Block Size in Bytes': '0x1000', 'Device Total Physical Paritions': '0x6', 'Device Manufacturer ID': '0x198', 'Device Serial Number':
'0x0', 'UFS fInitialized': '0x1', 'UFS Current LUN Number: ': ' 0xd0', 'UFS Total Active LU': '0x6', 'UFS wManufacturerID': '0x198', 'UFS Boot Partition Enabled': '0x1', 'UFS Raw Device Capacity:
': ' 0x3b9e000', 'UFS Min Block Size': '0x8', 'UFS Erase Block Size': '0x2000', 'UFS Allocation Unit Size': '0x1', 'UFS RPMB ReadWrite Size: ': ' 0x40', 'UFS Number of Allocation Uint for This LU'
: '0x0', 'UFS Logical Block Size': '0x0', 'UFS Provisioning Type': '0x0', 'UFS LU Write Protect': '0x0', 'UFS Boot LUN ID: ': ' 0x0', 'UFS Memory Type': '0x0', 'UFS LU Total Blocks': '0x0', 'UFS S
upported Memory Types': '0x801f', 'UFS dEnhanced1MaxNAllocU': '0x1dcf', 'UFS wEnhanced1CapAdjFac': '0x200', 'UFS dEnhanced2MaxNAllocU: ': ' 0x20', 'UFS wEnhanced2CapAdjFac': '0x200', 'UFS dEnhance
d3MaxNAllocU': '0x0', 'UFS wEnhanced3CapAdjFac': '0x0', 'UFS dEnhanced4MaxNAllocU': '0x0', 'UFS wEnhanced4CapAdjFac': '0x0', 'UFS LUN Enable Bitmask': '0x3f', 'UFS Logical Block Count': '0x0', 'UF
S bConfigDescrLock': '0x0', 'UFS iManufacturerName String Index': '0x1', 'UFS iProductName String Index': '0x2', 'UFS iSerialNumber String Index': '0x3', 'UFS iOemID String Index': '0x4', 'UFS Inq
uiry Command Output': 'TOSHIBA THGAF4G8N2LBAIRB0100 '}
firehose - {'Device Total Logical Blocks': '0x60ac00', 'Device Block Size in Bytes': '0x1000', 'Device Total Physical Paritions': '0x6', 'Device Manufacturer ID': '0x198', 'Device Serial Number':
'0x0', 'UFS fInitialized': '0x1', 'UFS Current LUN Number: ': ' 0xd0', 'UFS Total Active LU': '0x6', 'UFS wManufacturerID': '0x198', 'UFS Boot Partition Enabled': '0x1', 'UFS Raw Device Capacity:
': ' 0x3b9e000', 'UFS Min Block Size': '0x8', 'UFS Erase Block Size': '0x2000', 'UFS Allocation Unit Size': '0x1', 'UFS RPMB ReadWrite Size: ': ' 0x40', 'UFS Number of Allocation Uint for This LU'
: '0x0', 'UFS Logical Block Size': '0x0', 'UFS Provisioning Type': '0x0', 'UFS LU Write Protect': '0x0', 'UFS Boot LUN ID: ': ' 0x0', 'UFS Memory Type': '0x0', 'UFS LU Total Blocks': '0x0', 'UFS S
upported Memory Types': '0x801f', 'UFS dEnhanced1MaxNAllocU': '0x1dcf', 'UFS wEnhanced1CapAdjFac': '0x200', 'UFS dEnhanced2MaxNAllocU: ': ' 0x20', 'UFS wEnhanced2CapAdjFac': '0x200', 'UFS dEnhance
d3MaxNAllocU': '0x0', 'UFS wEnhanced3CapAdjFac': '0x0', 'UFS dEnhanced4MaxNAllocU': '0x0', 'UFS wEnhanced4CapAdjFac': '0x0', 'UFS LUN Enable Bitmask': '0x3f', 'UFS Logical Block Count': '0x0', 'UF
S bConfigDescrLock': '0x0', 'UFS iManufacturerName String Index': '0x1', 'UFS iProductName String Index': '0x2', 'UFS iSerialNumber String Index': '0x3', 'UFS iOemID String Index': '0x4', 'UFS Inq
uiry Command Output': 'TOSHIBA THGAF4G8N2LBAIRB0100 '}
firehose - 0x1
firehose_client - Supported functions:
-----------------
firehose -
Writing to physical partition 4, sector 6, sectors 128
Progress: |██████████████████████████████████████████████████| 100.0% Write (Sector 0x80 of 0x80, ) 7.04 MB/s
Wrote rpmbak.img to sector 6.
```

Hopefully you know how to fix that in a better way, or maybe this one will help the others to get through.

**Comment by bkerler** [comment]
I've added the fix to 3cf3324, it's not prince charming, but still does its job, thanks !
I would like to see a pr commit however next time, if possible.

**Comment by rabits** [comment]
Yeah, sorry, thought it's completely inappropriate and thought it will be a reference... Thank you anyway!
